internal/processmon/processmon.go

```diff
@@ -0,0 +1,379 @@
+	processMonitorCacheName = "ProcessMonitorCache"
+	secretLength            = 32
+)
+
+// processMon is an instance of processMonitor
+type processMon struct {
+	// netNSPath made configurable to enable running tests
+	netNSPath       string
+	activeProcesses *cache.Cache
+	childExitStatus chan exitStatus
+	// logToConsole stores if we should log to console
+	logToConsole bool
+	// launcProcessArgs are arguments that are provided to all processes launched by processmon
+	launcProcessArgs []string
+}
+
+// processInfo stores per process information
+type processInfo struct {
+	contextID string
+	RPCHdl    rpcwrapper.RPCClient
+	process   *os.Process
+}
+
+// exitStatus captures the exit status of a process
+type exitStatus struct {
+	process int
+	// The contextID is optional and is primarily used by remote enforcer
+	// processes to represent the namespace in which the process was running
+	contextID  string
+	exitStatus error
+}
+
+func init() {
+	// Setup new launcher
+	newProcessMon(netNSPath)
+}
+
+// contextID2SocketPath returns the socket path to use for a givent context
+func contextID2SocketPath(contextID string) string {
+
+	if contextID == "" {
+		panic("contextID is empty")
+	}
+	return filepath.Join("/var/run/", contextID+".sock")
+}
+
+// processIOReader will read from a reader and print it on the calling process
+func processIOReader(fd io.Reader, contextID string, exited chan int) {
+
+	reader := bufio.NewReader(fd)
+
+	for {
+		str, err := reader.ReadString('\n')
+		if err != nil {
+			exited <- 1
+			return
+		}
+
+		fmt.Print("[" + contextID + "]:" + str)
+	}
+}
+
+// newProcessMon is a method to create a new processmon
+func newProcessMon(netns string) ProcessManager {
+
+	launcher = &processMon{
+		netNSPath:       netns,
+		activeProcesses: cache.NewCache(processMonitorCacheName),
+		childExitStatus: make(chan exitStatus, 100),
+	}
+
+	go launcher.collectChildExitStatus()
+
+	return launcher
+}
+
+// GetProcessManagerHdl returns a process manager handle.
+func GetProcessManagerHdl() ProcessManager {
+
+	return launcher
+}
+
+// collectChildExitStatus is an async function which collects status for all launched child processes
+func (p *processMon) collectChildExitStatus() {
+
+	for {
+
+		es := <-p.childExitStatus
+		zap.L().Debug("Remote enforcer exited",
+			zap.String("nativeContextID", es.contextID),
+			zap.Int("pid", es.process),
+			zap.Error(es.exitStatus),
+		)
+	}
+}
+
+// SetupLogAndProcessArgs setups args that should be propagated to child processes
+func (p *processMon) SetupLogAndProcessArgs(logToConsole bool, args []string) {
+
+	p.logToConsole = logToConsole
+	p.launcProcessArgs = args
+}
+
+// KillProcess sends a rpc to the process to exit failing which it will kill the process
+func (p *processMon) KillProcess(contextID string) {
+
+	s, err := p.activeProcesses.Get(contextID)
+	if err != nil {
+		zap.L().Debug("Process already killed or never launched")
+		return
+	}
+	req := &rpcwrapper.Request{}
```

Comment:
skip a line.

Also

```
req := &rpcwrapper.Request{
    Payload: s.(*processInfo).process.Pid,
}

Reply:
and hopefully `s` must always be good to assume this

Reply:
done